File: exercicios/projeto-final/skeletons/studio-ghibli/flutter/pratica/lib/screens/list_screen.dart
import 'package:flutter/material.dart';
import '../api/ghibli_api.dart';
import '../models/film.dart';
import 'detail_screen.dart';

const Map<String, String> _directors = {
  'miyazaki': 'Hayao Miyazaki',
  'takahata': 'Isao Takahata',
};

class ListScreen extends StatefulWidget {
  const ListScreen({super.key});

  @override
  State<ListScreen> createState() => _ListScreenState();
}

class _ListScreenState extends State<ListScreen> {
  final GhibliApi _api = GhibliApi();

  List<FilmSummary> _all = [];
  bool _loading = true;
  String? _error;
  String _searchText = '';
  String? _selectedDirectorSlug;
  Set<String>? _titlesForDirector;

  @override
  void initState() {
    super.initState();
    _loadList();
  }

  Future<void> _loadList() async {
    // TODO 1 (feature 1 — lista): chamar `_api.fetchList()`, guardar o
    // resultado em `_all` e marcar `_loading = false` via setState. Tratar
    // erro com `_error` (mensagem) igual ao padrão do catch abaixo.
    setState(() {
      _loading = false;
    });
  }

  Future<void> _selectDirector(String? slug) async {
    // TODO 4 (feature 4 — categoria/filtro): quando `slug` não é null,
    // chamar `_api.fetchTitlesByDirector(_directors[slug]!)` e guardar em
    // `_titlesForDirector` (setState). Quando `slug` é null, limpar
    // `_titlesForDirector`.
    setState(() {
      _selectedDirectorSlug = slug;
    });
  }

  List<FilmSummary> get _filtered {
    // TODO 3 (feature 3 — busca) + TODO 4 (feature 4 — categoria): filtrar
    // `_all` por `_titlesForDirector` (quando não-nulo, manter só quem tem
    // `title` no set) e por `_searchText` (substring case-insensitive do
    // `title`) — igual à lógica já implementada no gabarito.
    return _all;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Studio Ghibli')),
      body: Semantics(
        identifier: 'item-list-screen',
        child: _buildBody(),
      ),
    );
  }

  Widget _buildBody() {
    if (_loading) {
      return const Center(child: CircularProgressIndicator());
    }
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.all(12),
          child: Semantics(
            identifier: 'search-input',
            child: TextField(
              decoration: const InputDecoration(
                labelText: 'Buscar filme',
                border: OutlineInputBorder(),
                prefixIcon: Icon(Icons.search),
              ),
              onChanged: (value) => setState(() => _searchText = value),
            ),
          ),
        ),
        SizedBox(
          height: 44,
          child: ListView(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 12),
            children: [
              _buildDirectorChip(null, 'Todos'),
              for (final slug in _directors.keys) _buildDirectorChip(slug, slug),
            ],
          ),
        ),
        const SizedBox(height: 8),
        Expanded(
          child: ListView.builder(
            itemCount: _filtered.length,
            itemBuilder: (context, index) {
              final film = _filtered[index];
              return Semantics(
                identifier: 'item-card-${film.id}',
                child: ListTile(
                  title: Text(film.title),
                  onTap: () {
                    Navigator.of(context).push(
                      MaterialPageRoute(
                        builder: (_) => DetailScreen(filmId: film.id),
                      ),
                    );
                  },
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildDirectorChip(String? slug, String label) {
    final selected = _selectedDirectorSlug == slug;
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 4),
      child: Semantics(
        identifier: slug == null ? 'category-chip-all' : 'category-chip-$slug',
        child: ChoiceChip(
          label: Text(capitalize(label)),
          selected: selected,
          onSelected: (_) => _selectDirector(slug),
        ),
      ),
    );
  }
}

File: exercicios/projeto-final/skeletons/studio-ghibli/kmp/pratica/androidApp/src/main/kotlin/com/puciec/ghiblikmp/ListScreen.kt
package com.puciec.ghiblikmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.ghiblikmp.data.GhibliApi
import com.puciec.ghiblikmp.data.FilmSummary
import com.puciec.ghiblikmp.data.capitalize

private val DIRECTORS = mapOf(
    "miyazaki" to "Hayao Miyazaki",
    "takahata" to "Isao Takahata",
)

@Composable
fun ListScreen(api: GhibliApi, onSelect: (String) -> Unit) {
    var all by remember { mutableStateOf<List<FilmSummary>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    var searchText by remember { mutableStateOf("") }
    var selectedDirectorSlug by remember { mutableStateOf<String?>(null) }
    var titlesForDirector by remember { mutableStateOf<Set<String>?>(null) }

    LaunchedEffect(Unit) {
        // TODO 1 (feature 1 — lista): chamar api.fetchList(), guardar em `all`.
        // Tratar erro em `error` (try/catch) e marcar `loading = false` no final.
        loading = false
    }

    // TODO 3 (feature 3 — busca) + TODO 4 (feature 4 — categoria): filtrar
    // `all` por `titlesForDirector` (quando não-nulo, `titles.contains(it.title)`)
    // e por `searchText` (substring case-insensitive do `title`).
    val filtered = all

    if (loading) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { CircularProgressIndicator() }
        return
    }
    if (error != null) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { Text(error!!) }
        return
    }

    Box(Modifier.fillMaxSize().testTag("item-list-screen")) {
        LazyColumn(Modifier.fillMaxSize()) {
            item {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    label = { Text("Buscar filme") },
                    modifier = Modifier.fillMaxWidth().padding(12.dp).testTag("search-input"),
                )
            }
            item {
                LazyRow(
                    Modifier.fillMaxWidth().padding(horizontal = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    item {
                        FilterChip(
                            selected = selectedDirectorSlug == null,
                            onClick = { selectedDirectorSlug = null; titlesForDirector = null },
                            label = { Text("Todos") },
                            modifier = Modifier.testTag("category-chip-all"),
                        )
                    }
                    items(DIRECTORS.keys.toList()) { slug ->
                        FilterChip(
                            selected = selectedDirectorSlug == slug,
                            onClick = { selectedDirectorSlug = slug },
                            label = { Text(capitalize(slug)) },
                            modifier = Modifier.testTag("category-chip-$slug"),
                        )
                    }
                }
            }
            items(filtered, key = { it.id }) { film ->
                Text(
                    text = film.title,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag("item-card-${film.id}")
                        .clickable { onSelect(film.id) }
                        .padding(16.dp),
                    style = MaterialTheme.typography.bodyLarge,
                )
            }
        }
    }

    LaunchedEffect(selectedDirectorSlug) {
        // TODO 4 (feature 4 — categoria/filtro): quando `selectedDirectorSlug`
        // não é null, chamar api.fetchTitlesByDirector(DIRECTORS[slug]!!) e
        // guardar em `titlesForDirector`.
    }
}
